File: 114. FlattenBinaryTreeToLinkedList/FlattenBinaryTreeToLinkedList(1).js
// Runtime: 88 ms, faster than 77.52% of JavaScript online submissions for Flatten Binary Tree to Linked List.
// Memory Usage: 38.8 MB, less than 69.32% of JavaScript online submissions for Flatten Binary Tree to Linked List.
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {void} Do not return anything, modify root in-place instead.
 */
var flatten = function(root) {
    if (root == null)
        return null
    
    var array = []
    
    var inner_flatten = function(root){
        if (root == null)
            return

        array.push(root)

        if (root.left != null)
            inner_flatten(root.left)
        
        if (root.right != null)
            inner_flatten(root.right)
        
    }
    inner_flatten(root)

    for (var i=0; i<array.length-1; i++){
        array[i].left = null;
        array[i].right = array[i+1];
    }
};

File: 114. FlattenBinaryTreeToLinkedList/FlattenBinaryTreeToLinkedList(1).cs
// Runtime: 116 ms, faster than 31.50% of C# online submissions for Flatten Binary Tree to Linked List.
// Memory Usage: 24.9 MB, less than 58.35% of C# online submissions for Flatten Binary Tree to Linked List.
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public void Flatten(TreeNode root) {
        if (root == null) 
            return;
        
        List<TreeNode> list = new List<TreeNode>();
        
        InnerFlatten(root, list);
        
        for (int i=0; i<list.Count-1; i++){
            list[i].left = null;
            list[i].right = list[i+1];
        }
    }
    public void InnerFlatten(TreeNode root, List<TreeNode> list){
        if (root == null)
            return;
        
        list.Add(root);
        
        if (root.left != null)
            InnerFlatten(root.left, list);
        if (root.right != null)
            InnerFlatten(root.right, list);
        
    }
}